ppcs/ppc0022-metaprogramming.md

```diff
@@ -0,0 +1,367 @@
+
+## Rationale
+
+(explain why the (following) proposed solution will solve it)
+
+## Specification
+
+As this PPC adds a number of functions all on an overall-similar theme, they are all collected into a new toplevel namespace. For now this namespace is called `meta::`; though as per the rest of the details in this PPC that is currently very much up for debate.
+
+Objects that are returned by these functions will be instances of some named package within the `meta::` namespace, though we do not yet specify exactly what they will be. Furthermore we make no statement about what actual container type of objects those actually are. Users should not attempt to interact with these objects except via the API specified here.
+
+The get-like functions come in pairs, consisting of a function named `get_...` and one called `can_...`. If the target item exists, then in either case it is returned. The behaviour differs when the item does not exist. The `get` variant will throw an exception, whereas the `can` variant will return `undef`. This allows the caller to decide which behaviour is best for the use-case.
+
+It is currently still an undecided question on what ought to be the correct behaviour for `add`-like functions when asked to create something that already exists, or the `remove`-like functions when asked to remove something that doesn't exist. It may be solved by something similar to above, where there are two differently-named functions that differ in that case. As yet there aren't any particularly compelling solutions.
+
+### Functions to operate on Packages
+
+These are functions to create, manipulate, and inspect packages themselves.
+
+#### `get_package`
+
+```perl
+$metapackage = meta::get_package($name);
+```
+
+Returns a meta-package object instance to represent the named package, if such a package exists. If not an exception is thrown.
+
+[[TODO: Decide whether "all packages exist", and thus this can never fail]]
+
+#### `add_package`
+
+```perl
+$metapackage = meta::add_package($name);
+```
+
+Creates a new package of the given name, and returns a meta-package object instance to represent it.
+
+#### `list_packages`
+
+```perl
+@metapackages = meta::list_packages($basename);
+```
+
+Returns a list of sub-packages within the given base package name, each as a meta-package object instance.
+
+### Functions to operate on References
+
+Some less common functions that may still be useful in some situations. These functions can be used to obtain meta-programming object instances in ways other than walking the symbol table for given names.
+
+#### `for_reference`
+
+```perl
+$metavar = meta::for_reference($ref);
+```
+
+Given a reference to a package variable or subroutine, returns a meta-object instance to represent it. If `$varref` is not a SCALAR, ARRAY, HASH or CODE reference, an exception is thrown.
```

Comment:
You allow get_symbol to return a GLOB, but don't allow a globref to be passed in here.

Reply:
Ahyes; might as well allow GLOBrefs too. Fixed in 387d25b